File: internal/api/handlers/tokens.go
package handlers

import (
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/rexec/rexec/internal/storage"
)

// TokenHandler handles API token operations
type TokenHandler struct {
	store *storage.PostgresStore
}

// NewTokenHandler creates a new token handler
func NewTokenHandler(store *storage.PostgresStore) *TokenHandler {
	return &TokenHandler{store: store}
}

// CreateTokenRequest represents a request to create a new API token
type CreateTokenRequest struct {
	Name      string   `json:"name" binding:"required"`
	Scopes    []string `json:"scopes"`
	ExpiresIn *int     `json:"expires_in"` // Days until expiration (optional)
}

// CreateToken creates a new API token
func (h *TokenHandler) CreateToken(c *gin.Context) {
	userID := c.GetString("userID")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	var req CreateTokenRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "name is required"})
		return
	}

	// Calculate expiration if provided
	var expiresAt *time.Time
	if req.ExpiresIn != nil && *req.ExpiresIn > 0 {
		exp := time.Now().AddDate(0, 0, *req.ExpiresIn)
		expiresAt = &exp
	}

	// Default scopes
	scopes := req.Scopes
	if len(scopes) == 0 {
		scopes = []string{"read", "write"}
	}

	token, plainToken, err := h.store.GenerateAPIToken(c.Request.Context(), userID, req.Name, scopes, expiresAt)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create token"})
		return
	}

	// Return the token with the plain text (only shown once!)
	c.JSON(http.StatusCreated, gin.H{
		"token":        plainToken,
		"id":           token.ID,
		"name":         token.Name,
		"token_prefix": token.TokenPrefix,
		"scopes":       token.Scopes,
		"expires_at":   token.ExpiresAt,
		"created_at":   token.CreatedAt,
		"message":      "Save this token now. You won't be able to see it again!",
	})
}

// ListTokens returns all tokens for the current user
func (h *TokenHandler) ListTokens(c *gin.Context) {
	userID := c.GetString("userID")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	tokens, err := h.store.GetAPITokensByUserID(c.Request.Context(), userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list tokens"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"tokens": tokens})
}

// RevokeToken revokes an API token
func (h *TokenHandler) RevokeToken(c *gin.Context) {
	userID := c.GetString("userID")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	tokenID := c.Param("id")
	if tokenID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "token id required"})
		return
	}

	err := h.store.RevokeAPIToken(c.Request.Context(), userID, tokenID)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			c.JSON(http.StatusNotFound, gin.H{"error": "token not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to revoke token"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "token revoked"})
}

// DeleteToken permanently deletes an API token
func (h *TokenHandler) DeleteToken(c *gin.Context) {
	userID := c.GetString("userID")
	if userID == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "unauthorized"})
		return
	}

	tokenID := c.Param("id")
	if tokenID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "token id required"})
		return
	}

	err := h.store.DeleteAPIToken(c.Request.Context(), userID, tokenID)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			c.JSON(http.StatusNotFound, gin.H{"error": "token not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete token"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "token deleted"})
}

// ValidateToken validates an API token (used by middleware or CLI)
func (h *TokenHandler) ValidateToken(c *gin.Context) {
	// Get token from Authorization header
	authHeader := c.GetHeader("Authorization")
	if authHeader == "" {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "authorization header required"})
		return
	}

	token := strings.TrimPrefix(authHeader, "Bearer ")
	if token == authHeader {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "invalid authorization format"})
		return
	}

	apiToken, err := h.store.ValidateAPIToken(c.Request.Context(), token)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": err.Error()})
		return
	}

	// Get user info
	user, err := h.store.GetUserByID(c.Request.Context(), apiToken.UserID)
	if err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "user not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"valid":    true,
		"user_id":  user.ID,
		"username": user.Username,
		"email":    user.Email,
		"tier":     user.Tier,
		"scopes":   apiToken.Scopes,
	})
}

File: web-ui/src/lib/stores/tokens.ts
import { writable, get } from 'svelte/store';
import { auth } from './auth';

export interface APIToken {
  id: string;
  name: string;
  token_prefix: string;
  scopes: string[];
  last_used_at?: string;
  expires_at?: string;
  created_at: string;
}

interface TokensState {
  tokens: APIToken[];
  loading: boolean;
  error: string | null;
}

const API_BASE = '/api';

function createTokensStore() {
  const { subscribe, set, update } = writable<TokensState>({
    tokens: [],
    loading: false,
    error: null,
  });

  function getAuthHeader(): HeadersInit {
    const authState = get(auth);
    if (authState.token) {
      return { Authorization: `Bearer ${authState.token}` };
    }
    return {};
  }

  return {
    subscribe,

    async fetchTokens(): Promise<void> {
      update(s => ({ ...s, loading: true, error: null }));
      try {
        const res = await fetch(`${API_BASE}/tokens`, {
          headers: getAuthHeader(),
        });
        if (!res.ok) throw new Error('Failed to fetch tokens');
        const data = await res.json();
        update(s => ({ ...s, tokens: data.tokens || [], loading: false }));
      } catch (err: any) {
        update(s => ({ ...s, error: err.message, loading: false }));
      }
    },

    async createToken(name: string, scopes: string[] = ['read', 'write'], expiresIn?: number): Promise<{ token: string; id: string } | null> {
      update(s => ({ ...s, loading: true, error: null }));
      try {
        const res = await fetch(`${API_BASE}/tokens`, {
          method: 'POST',
          headers: {
            ...getAuthHeader(),
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ name, scopes, expires_in: expiresIn }),
        });
        if (!res.ok) throw new Error('Failed to create token');
        const data = await res.json();
        
        // Refresh the tokens list
        await this.fetchTokens();
        
        return { token: data.token, id: data.id };
      } catch (err: any) {
        update(s => ({ ...s, error: err.message, loading: false }));
        return null;
      }
    },

    async revokeToken(tokenId: string): Promise<boolean> {
      update(s => ({ ...s, loading: true, error: null }));
      try {
        const res = await fetch(`${API_BASE}/tokens/${tokenId}`, {
          method: 'DELETE',
          headers: getAuthHeader(),
        });
        if (!res.ok) throw new Error('Failed to revoke token');
        
        // Remove from local state
        update(s => ({
          ...s,
          tokens: s.tokens.filter(t => t.id !== tokenId),
          loading: false,
        }));
        
        return true;
      } catch (err: any) {
        update(s => ({ ...s, error: err.message, loading: false }));
        return false;
      }
    },

    async deleteToken(tokenId: string): Promise<boolean> {
      update(s => ({ ...s, loading: true, error: null }));
      try {
        const res = await fetch(`${API_BASE}/tokens/${tokenId}/permanent`, {
          method: 'DELETE',
          headers: getAuthHeader(),
        });
        if (!res.ok) throw new Error('Failed to delete token');
        
        // Remove from local state
        update(s => ({
          ...s,
          tokens: s.tokens.filter(t => t.id !== tokenId),
          loading: false,
        }));
        
        return true;
      } catch (err: any) {
        update(s => ({ ...s, error: err.message, loading: false }));
        return false;
      }
    },

    reset() {
      set({ tokens: [], loading: false, error: null });
    },
  };
}

export const tokens = createTokensStore();
